src/pubkey.h

```diff
@@ -203,6 +203,25 @@ class CPubKey
     bool Derive(CPubKey& pubkeyChild, ChainCode &ccChild, unsigned int nChild, const ChainCode& cc) const;
 };
 
+class XOnlyPubKey {
+private:
+    uint256 m_keydata;
+
+public:
+    XOnlyPubKey(const uint256& in) : m_keydata(in) {}
+
+    /** Verify a 64-byte Schnorr signature.
+     *
+     * If the signature is not 64 bytes, or the public key is not fully valid, false is returned.
+     */
+    bool VerifySchnorr(const uint256& hash, const std::vector<unsigned char>& vchSig) const;
```

Comment:
It would be nice if there was an `IsValid()` function to check if the pubkey is actually valid, as opposed to the signature being invalid.

Reply:
There should be no need.

Public keys as specified by BIP340 are 32-byte arrays, not points or X coordinates or whatever. The signature validation algorithm takes as input a 32-byte public key, and 64-byte signature, and a 32-byte message.

This is intentionally different from how ECDSA works, where differences between encoding errors and invalid signatures are significant, and arguably the cause of all the trouble we went through with ECDSA encodings, and BIP66...

So, to be clear: there should never be a behavior difference between an invalid signature and invalidity due to invalid public keys. Both are just inputs for which the verification returns false. Having a way to distinguish the two would be a risk.

Reply:
> It would be nice if there was an `IsValid()` function to check if the pubkey is actually valid, as opposed to the signature being invalid.

```
class XOnlyPubKey {
private:
    uint256 m_keydata;

public:
    XOnlyPubKey(const uint256& in) : m_keydata(in) {}

    /** Verify a 64-byte Schnorr signature.
     *
     * If the signature is not 64 bytes, or the public key is not fully valid, false is returned.
     */
    bool VerifySchnorr(const uint256& hash, const std::vector<unsigned char>& vchSig) const;
    bool CheckPayToContract(const XOnlyPubKey& base, const uint256& hash, bool sign) const;

    const unsigned char& operator[](int pos) const { return *(m_keydata.begin() + pos); }
    const unsigned char* data() const { return m_keydata.begin(); }
    size_t size() const { return 32; }
};
```

This on-line instruction is already doing the job 
``XOnlyPubKey(const uint256& in) : m_keydata(in) {}``

Reply:
@sipa It is not useful to be able to distinguish between an invalid pubkey and an invalid signature? I don't understand the reasoning there. Indeed it should be helpful to be able to tell a user which part they fat fingered, if any.

Reply:
@StEvUgnIn No, I am talking about the call to secp256k1_xonly_pubkey_parse failing.

Reply:
Sure, for users it may be useful to have a sanity check. But consensus code should never distinguish between invalid public key and invalid signature.

Reply:
I agree. That said, having `IsValid()` would be helpful for non-consensus use.. I guess that would be tempting to use in consensus code, so a warning comment might be needed.

Reply:
@kallewoof OK

Reply:
@kallewoof So, I suggest that maybe we add a way to test if a pubkey is valid later when wallet support is added. But as the current PR is solely consensus/validation code, I think it's simpler and safer to not expose that distinction for now.